### ISSUE TITLE
Not declaring that a task uses a shared build service becomes an error in 10.0

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#22888]
Behavior deprecated in 7.4

* https://github.com/gradle/gradle/pull/19342

Note that this applies only with the `STABLE_CONFIGURATION_CACHE` feature flag enabled.

Kotlin Gradle Plugin still does not declare shared build service usages

* [KT-52149 - Gradle: declare shared build services usages with `Task#usesService`](https://youtrack.jetbrains.com/issue/KT-52149/Gradle-declare-shared-build-services-usages-with-TaskusesService)

**Comment by hungvietnguyen** [comment]
It is extremely easy for users to forget to call `Task.usesService`. In AGP, we have so many violations (https://issuetracker.google.com/300617088), and fixing them *exhaustively* is quite challenging.

**Is there a way Gradle could help users by calling `Task.usesService` automatically?** For example, after Gradle finishes a task's configuration, it can check all the task's properties to see if any properties are `RegisteredBuildServiceProvider`, and if so, automatically calls `Task.usesService` directly.

// cc @abstratt who implemented `TaskRequiredServices`

**Comment by hungvietnguyen** [comment]
*(I've asked the following on Gradle Slack, but perhaps it's good to post it here too.)*

It seems that `@ServiceReference` (without a service name) is a nicer option than `Task.usesService`, so we're planning to use it more in our code. But `@ServiceReference` is currently incubating, so I wanted to check on the following:
  1. When do you think `@ServiceReference` will be stable?
  2. Currently, Gradle does not accept `Property<out BuildService>` without an annotation. So users would normally have to annotate it with `@Internal`. Would it be nicer that Gradle provides *one of the following*?
      2a) Gradle requires users to annotate `Property<out BuildService>` with `@ServiceReference`.
      2b) [*stretch goal*] Gradle implicitly processes `Property<out BuildService>` as if it were already annotated with `@ServiceReference`, and therefore does not require users to annotate anything (i.e., the effect of adding `@ServiceReference` is exactly the same as not adding it).

**Comment by abstratt** [comment]
See: https://github.com/gradle/gradle/issues/30858

**Comment by hungvietnguyen** [comment]
Thanks, it would be great to get your thoughts on question 2 too.

**Comment by abstratt** [comment]
Personally, I tend to favor explicitness over magic, and for consistency, we should require explicit annotations (and that would be consistent with the other kinds of properties). Also, we should not forbid people using types that implement `BuildService` from being used as `@Internal`-annotated properties (as it has already been required with `usesService()`), at least not before we deprecate `usesService()`, if for whatever reason we ever choose to do that.

**Comment by hungvietnguyen** [comment]
// cc @liutikas 

I think the users also like "explicitness over magic", but this case is slightly different: When a Task has a `Property<out BuildService>`, it means that the task **will use** the build service. So requiring the association between the two seems a bit unnecessary.

In increasing order of difficulty, it seems that the options for the users are as follows:
   1. Users don't need to declare anything (i.e., Gradle removes the `Task.usesService` requirement).
   2. Users have a way to identify all places at compile time that they need to annotate `@ServiceReference` (i.e., Gradle validation checks will help the users with this). The Gradle checks must be complete, such that Gradle does not suddenly fail at run time requiring `Task.usesService`.
  3. Same as 2, but instead of getting help from Gradle, users need to **manually** identify all those places, which is tedious and easy to miss.
  4. Same as 3, but instead of annotating `@ServiceReference`, users need to manually add **`Task.usesService`**, which is even more tedious//easier to miss than `@ServiceReference`.

As you can see, the current available options are 3 and 4, which are not exactly great. Having options 1 or 2 would be really nice.

**Comment by abstratt** [comment]
I see now. This is an example of a validation error for a property with build service type that does not have an annotation:

```
A problem was found with the configuration of task ':named' (type 'Consumer').
  - Type 'Consumer' property 'counter' is missing an input or output annotation.
    
    Reason: A property without annotation isn't considered during up-to-date checking.
    
    Possible solutions:
      1. Add an input or output annotation.
      2. Mark it as @Internal.
```

I think option 2 is quite reasonable. I opened https://github.com/gradle/gradle/issues/30927 for that.

**Comment by hungvietnguyen** [comment]
That's great, thank you!

Just one more thought to consider: Is there a valid use case where a task has a build service property but **doesn't use** it?
   - If there is no use case, then the `Task.usesService` requirement is entirely unnecessary.
   - If there is a use case (please give a **concrete example**), it's likely just a user mistake. Gradle can still treat all build service properties **as if they would be used**, regardless of whether they will actually be used or not, and therefore also eliminate the `Task.usesService` requirement.

**Comment by abstratt** [comment]
> the Task.usesService requirement is entirely unnecessary.

For most scenarios, folks should use `@ServiceReference` and, in that case, `usesService` is not needed (as it is redundant).  That is what `@ServiceReference` was created for (https://github.com/gradle/gradle/issues/16170). Once `@ServiceReference` is taken out of incubation, it is possible we may in the future eventually deprecate `@Internal` and `usesService` for build service properties.

>  Is there a valid use case where a task has a build service property but doesn't use it?

A task could hold references to multiple services, but only use one of them based on some other configuration (for instance, a production service vs a development service), and so it would invoke `usesService` on the chosen service, but not on the other one (or assign it to a `@ServiceReference` property). In that case, automatically inferring a task uses a service just because it holds a reference to it it would be incorrect.

**Comment by hungvietnguyen** [comment]
Sorry I wasn't clear: By `Task.usesService` requirement, I meant the association between a Task and a service, whether by the `@ServiceReference` annotation or the `Task.usesService` API. I've updated my previous comment.

> A task could hold references to multiple services, but only use one of them based on some other configuration (for instance, a production service vs a development service)

Okay, IIUC your example is:
  - During task configuration: 
```
task.service1.set(foo)
task.service2.set(bar)
```
  - During task execution:
```
if (someCondition) {
   task.service1.get().doSomething()
} else {
   task.service2.get().doSomething()
}
```

To avoid the current Gradle errors for the task-service association, the users would have no choice but to fix the task configuration as follows:
```
task.service1.set(foo)
task.usesService(foo) // Or set @ServiceReference on service1
task.service2.set(bar)
task.usesService(bar) // Or set @ServiceReference on service2
```

That means whenever a task sets a property to a build service, the user would **always** need to associate the two. That means the requirement is unnecessary.

**Comment by abstratt** [comment]
The choice between which service to use would need to be done at *configuration* time:

At configuration time: 
```
if (someConfigTimeCondition) {
   serviceToUse = productionService
} else {
   serviceToUse = developmentService
}
// only if usesService were not a @ServiceReference:
usesService(serviceToUse)
```

At execution time:
```
serviceToUse.get().doSomething()
```

**Comment by hungvietnguyen** [comment]
*Maybe let's leave `@ServiceReference` aside to simplify our discussion for now.*

Your example can be simplified as:
```
someService = selectService() // The logic inside selectService() doesn't matter
task.serviceToUse = someService
task.usesService(someService)
```

That still means whenever a task sets a property to a build service, the user would **always** need to associate the two.

**Comment by abstratt** [comment]
You are assuming the task would only be assigned the selected service - the choice would be external. That is one possibility. But the choice of which service to use could be made elsewhere, such as in the task implementation itself, and the task would be fed with both options of services. 

> Maybe let's leave @ServiceReference aside to simplify our discussion for now.

Well, `@ServiceReference` is the solution we introduced in 8.0 to address the issue of requiring users to remember to call `usesService`, which seems to be your main concern here (while also bringing automatic injection of services by name/type), so not sure I understand the benefit of leaving it aside.

**Comment by hungvietnguyen** [comment]
Okay, let me sum up my key points:
  - The task-service association requirement is not necessary (option 1 in my [earlier comment](https://github.com/gradle/gradle/issues/22888#issuecomment-2414675239)). It is technically possible for Gradle to remove that requirement: For example, after Gradle finishes a task's configuration, it can check all the task's properties to see if any properties are `RegisteredBuildServiceProvider`, and if so, automatically calls `Task.usesService` directly without requiring the users to do it. Note that in this way, `Task.usesService` is entirely an internal Gradle concept, users don't need to know about it. => It's good for everyone.
  - If for some reason option 1 is not possible, then yes option 2 is the next best thing (https://github.com/gradle/gradle/issues/30927).

**Comment by abstratt** [comment]
Could you please enter a request for such feature?  However, I am guessing it probably would look a lot like https://github.com/gradle/gradle/issues/16170, for which `@ServiceReference` was devised.

**Comment by abstratt** [comment]
Also, assuming we can determine whether the consumed service does not specify `maxParallelUsages`, maybe we should not require usage to be explicitly declared, as AFAIK it brings no actual value.

**Comment by abstratt** [comment]
They would be related, yes, but I think both things could proceed in parallel. There may be cases we can automatically infer service usage, and there may be cases we can't. Your potential feature request could end up being addressed for the simpler (and most common) cases, so the error proposed here would be less likely to be triggered.  

At a minimum, we could have more focused discussions.  This issue is `a:chore` issue and, as such, at least typically, is not a good place for discussing alternative approaches, decisions have been made before/elsewhere, this issue is just a placeholder for the grunt work that will take care of it.

**Comment by hungvietnguyen** [comment]
Sure, I've filed [30989](https://github.com/gradle/gradle/issues/30989) - "Consider removing the Task.usesService requirement".
